In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression

df = pd.read_csv('demand_encoded.csv')

X = df.drop(columns=['demand'])
y = df['demand']

model = LinearRegression()
model.fit(X, y)

print("Model trained on full dataset.")
print(f"Features: {X.columns.tolist()}")

Model trained on full dataset.
Features: ['year', 'mineral_Cobalt', 'mineral_Copper', 'mineral_Lithium', 'mineral_Magnet rare earth elements', 'mineral_Nickel', 'technology_Electric vehicles', 'technology_Electricity networks', 'technology_Grid battery storage', 'technology_Hydrogen technologies', 'technology_Low emissions power generation', 'technology_Other low emissions power generation', 'technology_Solar PV', 'technology_Wind']


In [2]:
def predict_demand(mineral, technology, year):
    # start with a row of all zeros
    row = pd.DataFrame(columns=X.columns, data=[[0]*len(X.columns)])
    
    # set year
    row['year'] = year
    
    # set mineral flag
    mineral_col = f'mineral_{mineral}'
    if mineral_col in row.columns:
        row[mineral_col] = 1
    else:
        print(f"Unknown mineral: {mineral}")
        return None
    
    # set technology flag
    tech_col = f'technology_{technology}'
    if tech_col in row.columns:
        row[tech_col] = 1
    else:
        print(f"Unknown technology: {technology}")
        return None
    
    prediction = model.predict(row)[0]
    return round(prediction, 1)

In [3]:
print(predict_demand('Copper', 'Solar PV', 2030))
print(predict_demand('Lithium', 'Electric vehicles', 2035))
print(predict_demand('Nickel', 'Wind', 2040))
print(predict_demand('Cobalt', 'Electric vehicles', 2024))

1285.8
790.0
106.5
481.4


In [4]:
print(predict_demand('Copper', 'Electric vehicles', 2045))
print(predict_demand('Lithium', 'Electric vehicles', 2050))

2004.7
1243.9


## Step 6 Mineral demand predictor complete

A function that takes mineral + technology + year and returns a
demand forecast in kt.

### Sample predictions
- Copper, Solar PV, 2030 → 1285.8 kt
- Lithium, Electric vehicles, 2035 → 790.0 kt
- Copper, Electric vehicles, 2045 → 2004.7 kt  (extrapolated)
- Lithium, Electric vehicles, 2050 → 1243.9 kt  (extrapolated)

### Model: linear regression
- R²: 0.807
- Best performer across Step 1, 4, and 5
- 15 features: year + 5 minerals + 8 technologies (one-hot encoded)


In [5]:
print(predict_demand('Copper', 'Electric vehicles', 2055))
print(predict_demand('Lithium', 'Electric vehicles', 2055))

2307.3
1395.2
